### PR TITLE
Added networking event logger events

### PR DESCRIPTION
**Pull request by AMZN-Olex**
## What does this PR do?

Provides a high level stat system with Multiplayer gem (use MultiplayerStats.API target for the header).
Recommended to use the following macros. Here is an example of setting up a group and a few metric stats.
```cpp
    enum MultiplayerGroupIds
    {
        MultiplayerGroup_Networking
    };

    DECLARE_PERFORMANCE_STAT_GROUP(MultiplayerGroup_Networking, "Networking");
````
With a group defined, various stats can be defined to belong to a group.
```cpp
    enum MultiplayerStatIds
    {
        MultiplayerStat_EntityCount = 1000,
        MultiplayerStat_FrameTime
    };

    DECLARE_PERFORMANCE_STAT(MultiplayerGroup_Networking, MultiplayerStat_EntityCount, "NumEntities");
    DECLARE_PERFORMANCE_STAT(MultiplayerGroup_Networking, MultiplayerStat_FrameTime, "FrameTimeUs");
```

And then call SET_INTEGER_STAT to update stats as often as needed.

```cpp
    SET_PERFORMANCE_STAT(MultiplayerStat_EntityCount, m_entityCount);
    // ...
    SET_PERFORMANCE_STAT(MultiplayerStat_FrameTime, networkFrameTime);
```

Stats will be captured together within a group using AZ::EventLogger under the hood, which is configured using these cvars:
cl_metricsFile, sv_metricsFile and bg_enableNetworkingMetrics.

Example output into `C:\git\o3de-multiplayersample\user\Metrics\server_network_metrics.json`:
```
[
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618045780640,"pid":35824,"tid":7852,"args":{"NumEntities":0.0,"FrameTimeUs":27.7578125}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618046777584,"pid":35824,"tid":7852,"args":{"NumEntities":0.0,"FrameTimeUs":28.21875}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618047784584,"pid":35824,"tid":7852,"args":{"NumEntities":0.0,"FrameTimeUs":27.7109375}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618048862606,"pid":35824,"tid":7852,"args":{"NumEntities":2.0546875,"FrameTimeUs":31.390625}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618049791911,"pid":35824,"tid":7852,"args":{"NumEntities":55.1796875,"FrameTimeUs":83.75}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618050814717,"pid":35824,"tid":7852,"args":{"NumEntities":108.3046875,"FrameTimeUs":134.1171875}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618051795144,"pid":35824,"tid":7852,"args":{"NumEntities":148.6796875,"FrameTimeUs":179.421875}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618052811810,"pid":35824,"tid":7852,"args":{"NumEntities":193.3046875,"FrameTimeUs":243.2890625}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618053798298,"pid":35824,"tid":7852,"args":{"NumEntities":233.6796875,"FrameTimeUs":284.6484375}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618054798463,"pid":35824,"tid":7852,"args":{"NumEntities":272.0,"FrameTimeUs":320.75}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618055813071,"pid":35824,"tid":7852,"args":{"NumEntities":272.0,"FrameTimeUs":320.5546875}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618056809615,"pid":35824,"tid":7852,"args":{"NumEntities":272.0,"FrameTimeUs":319.0546875}},
  {"name":"Stats","cat":"Networking","ph":"C","ts":1665618057810645,"pid":35824,"tid":7852,"args":{"NumEntities":272.0,"FrameTimeUs":317.5546875}}
]
```

Note: current limitation, only UINT64 counters are supported while I hash out the public interface.

## How was this PR tested?

Using o3de-multiplayersample, dev branch, with Editor, GameLauncher, and ServerLauncher.

### CODE REVIEW

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 1]
```diff
@@ -1386,29 +1456,29 @@ namespace Multiplayer
...
-                blockLevelLoad = true;
+                    blockLevelLoad = true;
             }
             break;
         case MultiplayerAgentType::DedicatedServer:
             if (m_networkInterface->GetConnectionSet().GetConnectionCount() > 0)
             {
                 AZLOG_WARN("MultiplayerSystemComponent blocked this host from loading a new level because clients are connected. Loading a new level would destroy the existing clients' network player entity.")
-                blockLevelLoad = true;
+                    blockLevelLoad = true;
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:1481]: The identation is off here because the `AZLOG_WARN` macros are missing semicolons.
So Visual Studio adds an indent level when a previous line doesn't contains a semicolon

File: Gems/Multiplayer/Code/Source/MultiplayerStats.cpp [hunk 2]
```diff
@@ -253,4 +257,40 @@ namespace Multiplayer
...
+                m_framesSinceLastMetric = 0;
+                m_accumulatedNetworkTimeSinceLastMetric = AZ::Time::ZeroTimeUs;
+
+                AZ::Metrics::InstantArgs instantArgs;
+                instantArgs.m_name = "Metrics Event";
+                instantArgs.m_cat = "Networking";
+                instantArgs.m_args = argsContainer;
+                instantArgs.m_scope = AZ::Metrics::InstantEventScope::Process;
+
+                eventLogger->RecordInstantEvent(instantArgs);
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerStats.cpp:290]: This seems more like a Counter Event in according to the Google Trace Event Format [Document](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/edit#heading=h.msg3086636uq)

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 1 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp]: Good catch. Fixed.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 3]
```diff
@@ -842,15 +910,15 @@ namespace Multiplayer
         {
             AZLOG_INFO("New incoming connection from remote address: %s", connection->GetRemoteAddress().GetString().c_str())
 
-            MultiplayerAgentDatum datum;
+                MultiplayerAgentDatum datum;
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:913]: nitpick: indentation is also off here due to the AZLOG_INFO on the previous line

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 4]
```diff
@@ -416,6 +479,8 @@ namespace Multiplayer
     {
         AZ_PROFILE_SCOPE(MULTIPLAYER, "MultiplayerSystemComponent: OnTick");
 
+        const AZStd::chrono::utc_clock::time_point startMultiplayerTickTime = AZStd::chrono::utc_clock::now();
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:482]: This should be `steady_clock` not utc_clock.
The utc clock is affected by the system tiem and could go backwards if an NTP sync occurs(or the system clock is manually changed on the mahcine)

**Review by lumberyard-employee-dm — APPROVED** [review]

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 3 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp]: Fixed

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 4 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp]: Replaced with steady_clock

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 5]
```diff
@@ -508,6 +565,9 @@ namespace Multiplayer
             AZ_PROFILE_SCOPE(MULTIPLAYER, "MultiplayerSystemComponent: OnTick - SendReliablePackets");
             m_networkInterface->GetConnectionSet().VisitConnections(visitor);
         }
+
+        const auto duration = AZStd::chrono::duration_cast<AZStd::chrono::microseconds>(AZStd::chrono::steady_clock::now() - startMultiplayerTickTime);
+        stats.RecordFrameTime(AZ::TimeUs{ duration.count() });
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:570]: I think there should be type safe timer wrappers that will avoid you needing to cast between time types here.. AZ::ITime::GetRealElapsedTimeUs() and friends, although I admit the documentation is quite terrible here.. I have no idea what the difference is between all of RealSimulationTickTime, RealElapsedTime, ElapsedTime, ElapsedTimeDebug, SimulationTime, etc.. as an example; it looks like folks have been adding a kitchen sink worth of concepts without describing what they all are or why we need them.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 6]
```diff
@@ -1100,9 +1165,9 @@ namespace Multiplayer
     {
         IConnectionSet& connectionSet = m_networkInterface->GetConnectionSet();
         connectionSet.VisitConnections([readyForEntityUpdates](IConnection& connection)
-        {
-            connection.SendReliablePacket(MultiplayerPackets::ReadyForEntityUpdates(readyForEntityUpdates));
-        });
+            {
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:1168]: I think your editor is touching everything for some reason and making unrelated whitespace changes, which really clutters up your reviews and obfuscates your actual changes.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 7]
```diff
@@ -1220,21 +1285,21 @@ namespace Multiplayer
             AZStd::vector<NetBindComponent*> gatheredEntities;
             AZ::Interface<AzFramework::IVisibilitySystem>::Get()->GetDefaultVisibilityScene()->Enumerate(viewFrustum,
                 [this, &gatheredEntities](const AzFramework::IVisibilityScene::NodeData& nodeData)
-            {
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:1223]: This is a good example, it's really hard to tell if you actually touched anything in here without manually diffing almost line by line.

**Review by kberg-amzn — COMMENTED** [review]
Generally good changes.

I wonder if we're missing a layer of abstraction here however, just writing metrics into an event logger. The more usual pattern you'd find in other production games and engines is to have a metrics system where we write out global application stats just in memory, and then various metrics handlers that may sample the metrics and write out metrics files, or open a socket to monitoring dashboards like mrtg or munin or cloudwatch, or just dump metrics out to an ImGui window, or all of the above or whatever else a customer might want to do with metrics.. It feels a bit wrong to be doing stuff like having metrics file cvars and writing to an event logger system directly from simulation code like this. It means a lot of stuff will be decentralized, distributed, and duplicated that otherwise could have been centralized and far easier to manage.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:570]: We don't want any kind of simulation time here since the actual elapsed time is what needs to be measured.
It looks like that is [GetRealElapsedTimeUs](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Time/TimeSystem.cpp#L103-L110).

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 8]
```diff
@@ -134,59 +184,59 @@ namespace Multiplayer
                 ->Attribute(AZ::Script::Attributes::Module, "multiplayer")
                 ->Attribute(AZ::Script::Attributes::Category, "Multiplayer")
                 ->Method("GetOnEndpointDisconnectedEvent", [](AZ::EntityId id) -> EndpointDisconnectedEvent*
-                {
```
Comment (kberg-amzn) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:137]: Did you touch anything here? If you didn't, can you revert all these whitespace changes so it's easier to actually see what was touched?

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 5 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:570]: I wish we have a profile class that would measure this time for me.

Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:570]: Looking at:
``` cpp
    TimeUs TimeSystem::GetRealElapsedTimeUs() const
    {
        const TimeUs currentTime = static_cast<TimeUs>(AZStd::GetTimeNowMicroSecond());
        m_realAccumulatedTimeUs += currentTime - m_realLastInvokedTimeUs;
        m_realLastInvokedTimeUs = currentTime;

        return m_realAccumulatedTimeUs;
    }
```

This is not safe to call to get a valid results since if something calls in between start and end calls.

**Comment by lumberyard-employee-dm** [comment]
> Generally good changes.
> 
> I wonder if we're missing a layer of abstraction here however, just writing metrics into an event logger. The more usual pattern you'd find in other production games and engines is to have a metrics system where we write out global application stats just in memory, and then various metrics handlers that may sample the metrics and write out metrics files, or open a socket to monitoring dashboards like mrtg or munin or cloudwatch, or just dump metrics out to an ImGui window, or all of the above or whatever else a customer might want to do with metrics.. It feels a bit wrong to be doing stuff like having metrics file cvars and writing to an event logger system directly from simulation code like this. It means a lot of stuff will be decentralized, distributed, and duplicated that otherwise could have been centralized and far easier to manage.

The event logger API supports writing to anything that implements the [Generic Stream interface](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/GenericStreams.h#L20-L60).
Be that an in-memory stream(ByteContainerStream), file stream(SystemFileStream), network stream or even an az event stream.

An alternative API would be to have the event logger signal the event data through an `AZ::Event` in which the there can be multiple handlers a specific event. The problem with that approach is that we wouldn't want to signal the `AZ::Event` immediately synchronously, otherwise the connected handlers could be the bottleneck on logging of the event data, as they could perform operations synchronous operations which slows the actual metrics recording call.
If the event can't be signaled immediately then it gets exposes another problem of needing to queue the event data for later.
Depending on which kind of structure the data is stored in which can be dynamic(array) vs fixed (ring bugger), then that exposes another set of problems.
For a dynamic structure thread protection is now needed to protect that population of the buffer. Dynamic memory allocations also can cause a measure on the time when gather metrics which can pollute the sample size.
For a fixed structure thread protection is also needed. It doesn't have any issues with needing to allocate memory, but it is now vulnerable to initial issue as when it is full, it has to signal synchronously. In which case during certain frames when handlers are actually dealing with the metrics data could cause impact on the performance

Looking at an `AZ::ScheduleEvent` as another alternative that can't be used, as it can't queue any data, but only run a function at specified time intervals so that is not quite suited for using with the event logger.

The Task system can bind a [lambda with up to 56 bytes of data](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Task/Internal/Task.h#L113-L116) which could store a string instance in it(which would dynamically allocate the event data), but it is a message queue system and not a publish-subscribe so it can only be sent to a single sink

A best publish-subscribe system we have available, that can publish data in an asynchronous manner without blocking the current thread, is the EBus queue system, but it has it's own set of problems.
\+ It can queue the data for later dispatch and the queue isn't limited in size(dynamic structure)
\+ Adding more handlers doesn't slow the queuing operation
\- The queue can be dispatched by any code at any time as the [ExecuteQueuedEvents](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/EBus/BusImpl.h#L354-L368) function is a static public function on the EBus template, not part of some instant that can be hidden from the user
\- The event being queued has to be copied and results heap allocations
\- The EBus has to actually [queue an EBus event](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/EBus/BusImpl.h#L421-L428) that is part of an interface. This requires that handlers implement an EBus to handle the queued event.
\- The [EBus queued function API](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/EBus/BusImpl.h#L447-L462) of the EBus wouldn't work here as that models a [message queue](https://en.wikipedia.org/wiki/Message_queue) where in this case an asynchronous [publish-subscribe](https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern) pattern is needed to suite the use case of multiple metrics handlers.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:570]: The [duration begin and duration end](https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.nso4gcezn7n1) events are paired up and would measure the time for you.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 9]
```diff
@@ -96,6 +98,54 @@ namespace Multiplayer
...
+        }
+    }
+
+    void OnEnableNetworkingMetricsChanged(const bool& enabled)
+    {
+        UnregisterEventLoggerHelper();
+
+        if (enabled && GetMultiplayer())
+        {
+            switch (GetMultiplayer()->GetAgentType())
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:134]: Might want to have a WARN if the Agent Type if Uninitialized.

File: Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h [hunk 10]
```diff
@@ -93,5 +98,15 @@ namespace Multiplayer
         };
 
         void ConnectHandlers(EventHandlers& handlers);
+
+    private:
+        AZ::TimeUs m_accumulatedNetworkTimeSinceLastMetric = AZ::Time::ZeroTimeUs;
+        uint64_t m_framesSinceLastMetric = 0;
```
[the remark this answers is not in the document]
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h]: Sure, done

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 5 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:576]: You mean, that the processing of stream output would have to compute that? It seems easier to capture as one event for the first pass.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 6 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp]: Reverted whitespace changes. It should be good now.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 7 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp]: Reverted whitespace changes. It should be good now.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 8 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp]: Reverted whitespace changes. It should be good now.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 5 above)
Comment (lumberyard-employee-dm) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:576]: I mean there is two events for measuring a duration of time(the duration begin and end event) that should paired together.
There is also the complete event that combines the duration begin and end event into a single call, but the user is required to supply the duration in that case(https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/preview#heading=h.lpfof2aylapb)

**Comment by kberg-amzn** [comment]
> Out of all the alternatives, I think the best is to create a different implementation of the EventLogger interface, which is backed with a thread-safe event type

I'm just not sure, for global application metrics like we're talking about here, that an event logger interface is even the correct interface, I'm not sure why we'd want the intermediate form of this to be a stream of any sort to be honest. That seems like a downstream (hah!) concern from metrics gathering itself..  Here's some really simple stupid pseudocode that doesn't even come close to working, but hopefully give you a general idea. Obviously we'd need a general key/value store for metrics rather than a hard-coded struct, and our memory protection model would probably need to be better than a single reader/writer mutex, but hopefully the high level concepts make sense:

```
struct GlobalStats : public AZ::Interface
{
   AZ::TimeUs frameTimeUs = 0;
   AZ::TimeUs physicsTimeUs = 0;
   ...
   int memoryUsage = 0;
   int networkSendBytes = 0;
   ...
};

struct ScopedReadStats
{
   ScopedReadStats(GlobalStats* stats)
   {
      READERS_LOCK_ACQUIRE(mutex);
	  m_stats = stats;
   }
   
   ~ScopedReadStats()
   {
      m_stats = nullptr;
      READERS_LOCK_RELEASE(mutex);
   }
   
   const GlobalStats& operator->() const
   {
      return *stats;
   }

   GlobalStats* m_stats = nullptr;
};

void SetFrameTimeUs(AZ::TimeUs value)
{
   WRITERS_LOCK(mutex);
   GlobalStats.frameTimeUs = value;
}

void WriteStatsLog()
{
   // Write file stream containing current stats
   while (true)
   {
      ScopedReadStats readStats(GlobalStats);
	  // write readStats->values to stats.json
   }
}

void ImGuiStatsDisplay()
{
   ScopedReadStats readStats(GlobalStats);
   imgui.TextBox(readStats->FrameTimeUs);
   ...
}
```

etc.. So totally decouple the process of gathering and storing stats globally from an application standpoint from any particular output format. Anything that wants to stream this data can just sample it either on a time or on an edge trigger if we wanted to capture every single value change for some reason, and do its thing.

**Comment by lumberyard-employee-dm** [comment]
> > Out of all the alternatives, I think the best is to create a different implementation of the EventLogger interface, which is backed with a thread-safe event type
> 
> I'm just not sure, for global application metrics like we're talking about here, that an event logger interface is even the correct interface, I'm not sure why we'd want the intermediate form of this to be a stream of any sort to be honest. That seems like a downstream (hah!) concern from metrics gathering itself.. Here's some really simple stupid pseudocode that doesn't even come close to working, but hopefully give you a general idea. Obviously we'd need a general key/value store for metrics rather than a hard-coded struct, and our memory protection model would probably need to be better than a single reader/writer mutex, but hopefully the high level concepts make sense:
> 
> ```
> struct GlobalStats : public AZ::Interface
> {
>    AZ::TimeUs frameTimeUs = 0;
>    AZ::TimeUs physicsTimeUs = 0;
>    ...
>    int memoryUsage = 0;
>    int networkSendBytes = 0;
>    ...
> };
> 
> struct ScopedReadStats
> {
>    ScopedReadStats(GlobalStats* stats)
>    {
>       READERS_LOCK_ACQUIRE(mutex);
> 	  m_stats = stats;
>    }
>    
>    ~ScopedReadStats()
>    {
>       m_stats = nullptr;
>       READERS_LOCK_RELEASE(mutex);
>    }
>    
>    const GlobalStats& operator->() const
>    {
>       return *stats;
>    }
> 
>    GlobalStats* m_stats = nullptr;
> };
> 
> void SetFrameTimeUs(AZ::TimeUs value)
> {
>    WRITERS_LOCK(mutex);
>    GlobalStats.frameTimeUs = value;
> }
> 
> void WriteStatsLog()
> {
>    // Write file stream containing current stats
>    while (true)
>    {
>       ScopedReadStats readStats(GlobalStats);
> 	  // write readStats->values to stats.json
>    }
> }
> 
> void ImGuiStatsDisplay()
> {
>    ScopedReadStats readStats(GlobalStats);
>    imgui.TextBox(readStats->FrameTimeUs);
>    ...
> }
> ```
> 
> etc.. So totally decouple the process of gathering and storing stats globally from an application standpoint from any particular output format. Anything that wants to stream this data can just sample it either on a time or on an edge trigger if we wanted to capture every single value change for some reason, and do its thing.

This system isn't for global application metrics capturing.
It for per system metrics. The Networking metrics are separate from the rendering metrics, which are separate from the core metrics.

The Event Logger interface doesn't dictate that the metrics have to be immediately logged to a stream.
An implementation of the IEventLogger we have(JsonTraceEventLogger) sends the metrics directly to a stream which removes a lot of the concerns dealing with needing to store those statistics in memory and synchonize there readings across threads.

There is nothing preventing creating another implementation of the IEventLogger interface that stores statistics in a class such as the [StatisticalProfiler](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Statistics/StatisticalProfiler.h#L19-L41) and then write those statistics out at a later time.

However such system being suited for the multiplayer system needs, doesn't imply that it is suitable for other systems.

File: Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h [hunk 11]
```diff
@@ -20,11 +22,22 @@ namespace AzNetworking
...
 namespace Multiplayer
 {
+    enum MultiplayerGroupIds
+    {
+        MultiplayerGroup_Networking
+    };
+
+    enum MultiplayerStatIds
+    {
+        MultiplayerStat_EntityCount = 1000,
```
Comment (allisaurus) [inline Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h:32]: not clear to me why this is set to 1000?

Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h]: No reason, it can be any number but I like to start somewhere high enough to avoid collisions with other defined values of stat ids. For example, if we were to add them in o3de-multiplayersample

File: Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp [hunk 12]
```diff
@@ -105,12 +129,12 @@ namespace Multiplayer
             case MultiplayerAgentType::Client:
                 ConfigureEventLoggerHelper(cl_metricsFile);
                 break;
-            default:
-                AZLOG_WARN("Unitialized or unsupported agent type for recording metrics.");
+            case MultiplayerAgentType::Uninitialized:
+                AZLOG_WARN("Unitialized agent type isn't supported networking metrics.");
```
Comment (AMZN-puvvadar) [inline Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp:133]:
```suggestion
                AZLOG_WARN("Unitialized agent type isn't supported for networking metrics.");
```

File: Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerConstants.h [hunk 13]
```diff
@@ -9,11 +9,12 @@
 #pragma once
 
 #include <AzCore/EBus/Event.h>
+#include <AzCore/Metrics/IEventLoggerFactory.h>
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerConstants.h:12]: This file can forward declared `enum class AZ::Metrics::EventLoggerId : AZ::u32;` and avoid the need to include the IEventLogger.h

The IEventLogerFactory.h isn't actually needed either as it is IEventLogger.h that provides the EventLoggerId enum

File: Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp [hunk 14]
```diff
@@ -0,0 +1,238 @@
...
+                {
+                    AZStd::vector<AZ::Metrics::EventField> argsContainer;
+
+                    AZStd::for_each(
+                        group.m_stats.m_items.begin(),
+                        group.m_stats.m_items.end(),
+                        [&argsContainer](const CumulativeAverage& stat)
+                        {
+                            argsContainer.emplace_back(stat.m_name.c_str(), stat.m_average.CalculateAverage());
+                        });
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp:226]: You could use the [from_range](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/containers/vector.h#L145-L150) constructor to directly construct the argument container(more info about range 
```suggestion
                     auto TransformAverageToEventArg = [](const CumulativeAverage& stat)
                     {
                         return EventField{ stat.m_name, stat.m_average.CalculateAverage() };
                     };
                    AZStd::vector<AZ::Metrics::EventField> argsContainer(AZStd::from_range, group.m_stats.m_items | AZStd::views::transform(TransformAverageToEventArg));
```

File: Gems/Multiplayer/Code/Source/MultiplayerStats.cpp [hunk 15]
```diff
@@ -6,6 +6,8 @@
  *
  */
 
+#include <AzCore/Metrics/IEventLogger.h>
+#include <AzCore/Metrics/IEventLoggerFactory.h>
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerStats.cpp:10]: These includes aren't used in this file.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp [hunk 16]
```diff
@@ -28,6 +28,7 @@
 #include <AzCore/Console/IConsole.h>
 #include <AzCore/Console/ILogger.h>
 #include <AzCore/Math/ShapeIntersection.h>
+#include <AzCore/Metrics/JsonTraceEventLogger.h>
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp:31]: This include seems unused as well

File: Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h [hunk 17]
```diff
@@ -8,10 +8,12 @@
...
 #pragma once
 
-#include <AzCore/Time/ITime.h>
-#include <AzCore/std/containers/vector.h>
+#include <AzCore/EBus/ScheduledEvent.h>
 #include <AzCore/std/containers/array.h>
+#include <AzCore/std/containers/vector.h>
+#include <AzCore/Time/ITime.h>
 #include <Multiplayer/MultiplayerTypes.h>
+#include <Multiplayer/MultiplayerStatSystem.h>
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h:16]: The `ScheduleEvent.h`, `ITime.h` and `MultiplayerStatsSystem.h` includes don't look used in this file. Are they?

File: Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp [hunk 18]
```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Source/MultiplayerStatSystemComponent.h>
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp:9]: You can include `#include "MultiplayerStatSystemComponent"` to avoid needing to mention the Source folder in the #include

File: Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+        "server_network_metrics.json",
+        nullptr,
+        AZ::ConsoleFunctorFlags::DontReplicate,
+        "File of the server metrics file if enabled, placed under <ProjectFolder>/user/metrics");
+
+    void ConfigureEventLoggerHelper(const AZ::CVarFixedString& filename)
+    {
+        if (auto eventLoggerFactory = AZ::Interface<AZ::Metrics::IEventLoggerFactory>::Get())
+        {
+            const AZ::IO::FixedMaxPath metricsFilepath = AZ::IO::FixedMaxPath(AZ::Utils::GetProjectPath()) / "user/Metrics" / filename;
```
Comment (lumberyard-employee-dm) [inline Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp:48]: Should these metrics filenames add a timestamp to their name or are you looking into going with a rotating log system?

Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp:48]: I'll work on adding later if it is needed by AWSI.

**Review by galibzon — APPROVED** [review]

File: Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h (hunk 17 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerStats.h]: I've cleaned this up.

File: Gems/Multiplayer/Code/Source/MultiplayerStats.cpp (hunk 15 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerStats.cpp]: Cleaned up.

File: Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp (hunk 16 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerSystemComponent.cpp]: Cleaned up.

File: Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerConstants.h (hunk 13 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Include/Multiplayer/MultiplayerConstants.h]: Done

File: Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp (hunk 14 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp:224]: I'd rather avoid range API for now. But I'll keep it in mind.

File: Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp (hunk 18 above)
Comment (AMZN-Olex) [reply Gems/Multiplayer/Code/Source/MultiplayerStatSystemComponent.cpp:11]: I've been keeping everything as <> for consistency.

**Review by lumberyard-employee-dm — APPROVED** [review]
